date-time/src/commonMain/kotlin/com/splendo/kaluga/datetime/timer/RecurringTimer.kt

```diff
@@ -0,0 +1,190 @@
+            takeAndChangeState { state ->
+                when (state) {
+                    is State.Running -> state::pause
+                    is State.NotRunning -> state.remain()
+                }
+            }
+        }
+    }
+
+    private suspend fun finish() {
+        withContext(coroutineScope.coroutineContext) {
+            takeAndChangeState { state ->
+                when (state) {
+                    is State.NotRunning.Finished -> state.remain()
+                    is State.Running -> state::finish
+                    is State.NotRunning.Paused -> state::finish
+                }
+            }
+        }
+    }
+
+    /** Timer state. */
+    sealed class State : KalugaState(), Timer.State {
+        abstract val totalDuration: Duration
+        /** Timer is not running. */
+        sealed class NotRunning(override val elapsed: Duration) : State(), Timer.State.NotRunning {
+            /** Timer is paused. */
+            class Paused(
+                elapsedSoFar: Duration,
+                override val totalDuration: Duration
+            ) : NotRunning(elapsedSoFar), Timer.State.NotRunning.Paused {
+                fun start(
+                    interval: Duration,
+                    coroutineScope: CoroutineScope,
+                    finishCallback: suspend () -> Unit
+                ): Running = Running(
+                    elapsedSoFar = elapsed,
+                    totalDuration = totalDuration,
+                    interval = interval,
+                    coroutineScope = coroutineScope,
+                    finishCallback = finishCallback
+                )
+                internal fun finish(): Finished = Finished(totalDuration)
+            }
+
+            /** Timer is finished. */
+            class Finished(
+                override val totalDuration: Duration
+            ) : NotRunning(totalDuration), Timer.State.NotRunning.Finished
+        }
+
+        /** Timer is running. */
+        class Running(
+            elapsedSoFar: Duration,
+            override val totalDuration: Duration,
+            interval: Duration,
+            private val coroutineScope: CoroutineScope,
+            private val finishCallback: suspend () -> Unit
+        ) : State(), Timer.State.Running, HandleAfterNewStateIsSet<State>, HandleBeforeOldStateIsRemoved<State> {
+            override val elapsed: Flow<Duration> = tickProvider(
+                offset = elapsedSoFar,
+                max = totalDuration,
+                interval = interval
+            )
+            private lateinit var finishJob: Job
+
+            internal suspend fun pause(): NotRunning.Paused {
+                return NotRunning.Paused(elapsed.first(), totalDuration)
+            }
+            internal fun finish(): NotRunning.Finished = NotRunning.Finished(totalDuration)
+
+            override suspend fun beforeOldStateIsRemoved(oldState: State) {
+                finishJob = coroutineScope.launch {
```

Comment:
the Job idea is ok with me in principle, an alternate is to pass something into `tickProvider` (as you already break out of the loop there), this would create a single source or mockable time. Maybe you already considered the drawbacks of this approach though.

Also due to `late init var` usage this is not freezable, it's easy to fix by using a supervisor that uses `cancelChilderen` (has a similar discussion with @Daeda88 on this recently, maybe he has more insights on this).

Reply:
Yeah, good point. I usually pass some manager through the states and have the state transition methods start/stop the manager. But if you're going to use the state itself, make it like so:

private val superVisor = SupervisorJob()
            override suspend fun beforeOldStateIsRemoved(oldState: State) {
                CoroutineScope(superVisor + coroutineScope.coroutineContext).launch {
                   // DO stuff
                }
                
           override suspend fun afterNewStateIsSet(newState: State) {
               superVisor.cancelChildren()
           }
}

Reply:
done